lib/SILOptimizer/Mandatory/AddressLowering.cpp

```diff
@@ -596,944 +1198,1991 @@ class AddressMaterialization {
   AddressMaterialization(AddressLoweringState &pass, SILBuilder &B)
       : pass(pass), B(B) {}
 
-  SILValue initializeOperandMem(Operand *operand);
+  /// Return the address of the storage for `origValue`. This may involve
+  /// materializing projections. Record the materialized address as storage for
+  /// origValue. Called once at the definition of \p origValue.
+  SILValue materializeAddress(SILValue origValue) {
+    ValueStorage &storage = pass.valueStorageMap.getStorage(origValue);
+    if (storage.storageAddress)
+      return storage.storageAddress;
+
+    if (storage.isUseProjection) {
+      materializeUseProjectionStorage(storage, /*intoPhiOperand*/ false);
+    } else {
+      assert(storage.isDefProjection);
+      storage.storageAddress = materializeDefProjection(origValue);
+    }
+    return storage.storageAddress;
+  }
+
+  void initializeOperand(Operand *operand);
 
-  SILValue materializeAddress(SILValue origValue);
+  SILValue materializeUseProjectionStorage(ValueStorage &storage,
+                                           bool intoPhiOperand);
+
+  SILValue materializeDefProjection(SILValue origValue);
 
 protected:
-  SILValue materializeProjection(Operand *operand);
+  SILValue materializeStructExtract(SILInstruction *extractInst,
+                                    SILValue elementValue, unsigned fieldIdx);
+
+  SILValue materializeTupleExtract(SILInstruction *extractInst,
+                                   SILValue elementValue, unsigned fieldIdx);
+
+  SILValue materializeProjectionIntoUse(Operand *operand, bool intoPhiOperand);
+
+  SILValue materializeComposingUser(SingleValueInstruction *user,
+                                    bool intoPhiOperand) {
+    return materializeUseProjectionStorage(
+        pass.valueStorageMap.getStorage(user), intoPhiOperand);
+  }
 };
 } // anonymous namespace
 
-// Materialize an address pointing to initialized memory for this operand,
-// generating a projection and copy if needed.
-SILValue AddressMaterialization::initializeOperandMem(Operand *operand) {
+/// Given the operand of an aggregate instruction (struct, tuple, enum), ensure
+/// that the in-memory subobject is initialized. Generates an address
+/// projection and copy if needed.
+///
+/// If the operand projects into its use, then the memory was already
+/// initialized when visiting the use.
+void AddressMaterialization::initializeOperand(Operand *operand) {
   SILValue def = operand->get();
-  SILValue destAddr;
-  if (operand->get()->getType().isAddressOnly(*pass.F)) {
+  if (def->getType().isAddressOnly(*pass.function)) {
```

Comment:
Shouldn't this just be `def->getType().isAddress()`?
If it's an address type but not an address-only type we are creating a store here, which should be wrong.

Reply:
> Shouldn't this just be def->getType().isAddress()?

No. An address can't be passed into a struct/tuple. I renamed this to `initializeComposingUse`.